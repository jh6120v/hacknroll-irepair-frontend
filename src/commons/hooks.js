import React, { useState, useCallback, useEffect } from 'react';
import {
    useObservable,
    pluckFirst,
    useObservableState, useSubscription
} from 'observable-hooks';
import {
    map,
    filter,
    distinctUntilChanged,
    switchMap,
    repeat,
    withLatestFrom,
    scan,
    tap,
    takeWhile, take
} from 'rxjs/operators';
import { of, animationFrameScheduler, interval } from 'rxjs';
import { useDispatch } from 'react-redux';
import {
    Button, ModelContent, ModelFooter, ModelShadow, ModelTitle, ModelWrap
} from '../styles/model-style';
import { messagePush } from '../routes/chat/modules/chat-message';

const useModel = (title, message = null, confirm = null, confirmText = 'Confirm', cancel = null, cancelText = 'Cancel') => {
    const [isShown, setShown] = useState(false);
    const showModal = useCallback(() => setShown(true), []);
    const hideModal = useCallback(() => setShown(false), []);

    const ModelBox = () => (
        <div>
            <ModelWrap>
                <ModelTitle>{title}</ModelTitle>
                <ModelContent>{message}</ModelContent>
                <ModelFooter>
                    {
                        cancel ? (<Button type="cancel" onClick={cancel}>{cancelText}</Button>) : null
                    }
                    {
                        confirm ? (<Button type="confirm" onClick={confirm}>{confirmText}</Button>) : null
                    }
                </ModelFooter>
            </ModelWrap>
            <ModelShadow />
        </div>
    );

    return {
        ModelBox,
        isShown,
        showModal,
        hideModal
    };
};

const useMockData = (mockData, scrollBottom) => {
    const dispatch = useDispatch();
    const [mockState, setMockState] = useState('yet');

    const mockState$ = useObservable(pluckFirst, [mockState]);
    const timer$ = useObservable(() => interval(1000).pipe(
        take(4)
    ));

    const countDown$ = useObservable(() => mockState$.pipe(
        map((state) => state === 'yet'),
        distinctUntilChanged(),
        switchMap((isYet) => (isYet ? of(0) : timer$.pipe(
            tap((x) => {
                dispatch(messagePush({
                    singleMessage: mockData[x]
                }));

                scrollBottom();
            })
        )))
    ));

    useSubscription(countDown$);

    return {
        mockState,
        setMockState
    };
};

const useWebSocket = (host, { onOpen, onMessage, onClose, onError, reconnectInterval }) => {
    let connectInterval;
    let timeout = reconnectInterval || 250;

    const [ws, setWs] = useState(null);

    useEffect(() => {
        connect();
    }, []);

    const connect = () => {
        // wss://dxl9ub4w15.execute-api.us-west-2.amazonaws.com/stage
        const wss = new WebSocket(host);

        wss.onopen = (e) => {
            console.log("on open.");

            setWs(wss);

            onOpen(e);

            setTimeout(() => {
                wss.send('123');
            }, 3000);
        };

        wss.onmessage = (e) => {
            let received_msg = JSON.parse(e.data);

            onMessage(received_msg, e);
        };

        wss.onclose = (e) => {
            onClose(e);

            console.log(
                `Socket is closed. Reconnect will be attempted in ${Math.min(
                    10000 / 1000,
                    (timeout + timeout) / 1000
                )} second.`,
                e.reason
            );

            timeout = timeout + timeout;
            connectInterval = setTimeout(check, Math.min(10000, timeout));
        };

        wss.onerror = err => {
            onError(err);

            console.error(
                "Socket encountered error: ",
                err.message,
                "Closing socket"
            );

            wss.close();
        };

        const check = () => {
            // check if websocket instance is closed, if so call `connect` function.
            if (!wss || wss.readyState === WebSocket.CLOSED) {
                connect();
            }
        };
    };

    return ws;
};

const useTimer = (initialTime = 0, direction = 'forward', onSuccess) => {
    const [timerState, setTimerState] = useState('reset');

    const timerState$ = useObservable(pluckFirst, [timerState]);
    const countDown$ = useObservable(() => timerState$.pipe(
        map((state) => state === 'reset'),
        distinctUntilChanged(),
        switchMap((isReset) => (isReset ? of(initialTime)
            // high accuracy timing
            // repetitively calculate the diff
            : of(animationFrameScheduler.now(), animationFrameScheduler).pipe(
                repeat(),
                // extract seconds
                map((startTime) => ~~((Date.now() - startTime) / 1000)),
                distinctUntilChanged(),
                withLatestFrom(timerState$),
                filter(([, state]) => state === 'started'),
                tap(([val,]) => {
                    // console.log(`tab::${val}`);

                    if (direction === 'backward' && val > initialTime) {
                        onSuccess();
                    }
                }),
                takeWhile(([val,]) => direction === 'forward' || (direction === 'backward' && val <= initialTime)),
                scan((timeLeft, [val,]) => {
                    // count how many second left

                    // 當時間尚未經過1秒時，先不進行時間增減
                    if (val === 0) {
                        return timeLeft;
                    }

                    return direction === 'backward' ? timeLeft - 1 : timeLeft + 1;
                }, initialTime)
            )))
    ));

    const seconds = useObservableState(countDown$, initialTime);

    return {
        timerState,
        setTimerState,
        seconds
    };
};

export {
    useModel,
    useTimer,
    useMockData,
    useWebSocket
};
