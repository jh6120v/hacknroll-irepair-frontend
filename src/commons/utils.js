import React, { Suspense } from 'react';
import moment from 'moment';
import { LazyLoad } from '../styles/common-style';
import { WEBSOCKET_URL } from '../constants';

export const createActionCreator = (namespace) => (actionType) => {
    const type = `${namespace}/${actionType}`;
    const actionCreator = (payload = {}) => ({
        type,
        payload
    });

    actionCreator.type = type;
    Object.freeze(actionCreator);

    return actionCreator;
};

export const waitingRouteComponent = (Component) => () => (
    <Suspense fallback={<LazyLoad>Loading...</LazyLoad>}>
        <Component />
    </Suspense>
);

export const getRandom = (min, max) => Math.floor(Math.random() * (max - min + 1)) + min;

export const getRandomID = () => {
    return `A${+moment()}`;
};

export const getRandomAvatar = () => {
    const avatarList = [
        'avatar-1.png',
        'avatar-2.png',
        'avatar-3.png',
        'avatar-4.png',
        'avatar-5.png',
        'avatar-6.png'
    ];

    const key = getRandom(0, 5);

    return avatarList[key];
};

export const getWebSocketUrl = (roomId = '') => WEBSOCKET_URL + roomId;

export const getRandomAuthor = () => {
    const authorList = [
        'Cosmo',
        'Hunk',
        'Jerome',
        'Gary',
        'Phoebe',
        'Rock',
        'Spark',
        'Vincent',
        'Bess',
        'Ella',
        'Cherry',
        'Flora',
        'Ivy',
        'Kelly',
        'Maggie',
        'Nicole',
        'Sami',
        'Polly',
        'Vera',
        'Zora'
    ];

    const key = getRandom(0, 19);

    return authorList[key];
};
