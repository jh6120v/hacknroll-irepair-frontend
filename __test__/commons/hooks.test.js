import React from 'react';
import { renderHook, act } from '@testing-library/react-hooks';
import { render } from '@testing-library/react';
import { useModel, useTimer } from '../../src/commons/hooks';

describe('test custom hooks', () => {
    test('test useModel hooks', () => {
        const { result } = renderHook(() => useModel(
            'test',
            'message',
            () => {
                console.log('confirm test');
            },
            'confirm',
            () => {
                console.log('cancel test');
            },
            'cancel'
        ));

        const { container, getByText } = render(<result.current.ModelBox />);

        expect(container.firstChild).toMatchSnapshot();

        expect(getByText('confirm')).toBeTruthy();
        expect(getByText('cancel')).toBeTruthy();

        act(() => {
            result.current.showModal();
        });

        expect(result.current.isShown).toBe(true);

        act(() => {
            result.current.hideModal();
        });

        expect(result.current.isShown).toBe(false);
    });

    it('test useTimer hooks', () => {
        const { result } = renderHook(() => useTimer(3, 'backward'));

        expect(result.current.timerState).toEqual('reset');

        act(() => {
            result.current.setTimerState('start');
        });

        expect(result.current.timerState).toEqual('start');
    });
});
