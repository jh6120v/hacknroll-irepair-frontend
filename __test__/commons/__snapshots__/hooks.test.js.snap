// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test custom hooks test useModel hooks 1`] = `
<div>
  <div
    class="sc-bwzfXH ebVDIy"
  >
    <div
      class="sc-htpNat uEuno"
    >
      test
    </div>
    <div
      class="sc-bxivhb iCCdRw"
    >
      message
    </div>
    <div
      class="sc-ifAKCX kdvBBw"
    >
      <div
        class="sc-EHOje kNhJuu"
        type="cancel"
      >
        cancel
      </div>
      <div
        class="sc-EHOje brQrel"
        type="confirm"
      >
        confirm
      </div>
    </div>
  </div>
  <div
    class="sc-bdVaJa giODjc"
  />
</div>
`;
